Add TF ViT MAE

sayakpaul opened the pull request:
This PR adds the MAE [1] model in TensorFlow. It was developed by @arig23498 and myself. 

Fun facts about this PR:

* Probably the third pure vision model in TensorFlow in `transformers`.

**References**:

[1] [Masked Autoencoders Are Scalable Vision Learners](https://arxiv.org/abs/2111.06377)

**Update**

The PR is now ready for review. @gante @Rocketknight1 @sgugger

sayakpaul committed:
partially ported pt methods and classes of vit mae to tensorflow. (#1)

* partially ported pt methods and classes of vit mae to tensorflow.

* ported TFViTMAEIntermediate and TFViTMAEOutput.

* chore: addresses PR feedback.

* added TFViTMAEModel and started TFViTMAEDecoder.

* add: initial implementation of tf vit mae.

* fix: model output type.

* fix: a bunch of inconsistencies but need to investigate tf.repeat().

* chore: pr feedback.

* fix: gather error.

* chore: resorted to the original pt vit_mae model.

* fix: fix gather error partially (ral this time).

* feat: adding tf vit mae; model initializing but weight porting is flawed.

* partial fix: fixing the parameter names

Fixed the parameter names of dropout layers, decoder_embed and the decoder_layers. This helps with the proper cross-loading of the weights. With proper cross loading the debug test does not pass. My intuition is to dive a little deeper into the computations of the model

* chore: applied make style.

Co-authored-by: ariG23498 <aritra.born2fly@gmail.com>

ariG23498 committed:
Merge branch 'master' into tf-mae
OK

ariG23498 committed:
Merge branch 'master' into tf-mae
OK

ariG23498 committed:
Merge branch 'master' into tf-mae
OK

ariG23498 committed:
Merge branch 'master' into tf-mae
OK

sayakpaul committed:
feat: monkey patching of tf.random.uniform to control randomness.

sayakpaul committed:
feat: added a noise argument in the implementation for reproducibility.

sayakpaul committed:
feat: vit mae models with an additional noise argument for reproducibility.

ydshieh commented on src/transformers/models/vit_mae/modeling_vit_mae.py:
```diff
@@ -242,18 +245,21 @@ def initialize_weights(self):
         Args:
             sequence (`torch.LongTensor` of shape `(batch_size, sequence_length, dim)`)
+            noise (`tf.Tensor` of shape `(1, sequence_length)`) which is only used
+                for testing purposes to control randomness
```
Here should be `torch.FloatTensor`. The shape is still better to stated as `(batch_size, sequence_length)`.
Although we introduce it for testing purpose, I would suggest to say:

... which is **mainly** used for testing purposes to control randomness and maintain the reproducibility

sayakpaul replied:
Will take care of that.

ydshieh commented:
I just have a quick look and left a comment.

I feel strange that there are style changes like `omega = 1.0 / 10000 ** omega # (D/2,)`. Do you have a previous version of `black`?

sayakpaul commented:
> I just have a quick look and left a comment.
> 
> I feel strange that there are style changes like `omega = 1.0 / 10000 ** omega # (D/2,)`. Do you have a previous version of `black`?
> 
> I thought you already updated the version (during your work on `TFConvNext`), via
> 
> ```
> pip install -e .[quality]
> ```
> 
> Maybe you were in a different virtual Python environment while working on this PR?

So, first I deactivate the current Python virtual environment and then run the installation and run `make style`? 

I think I should fetch upstream before that and rebase.

ydshieh commented:
> So, first I deactivate the current Python virtual environment and then run the installation and run `make style`?

Yeah, if your current venv is specific to your other work/projects, and you don't want to change its installed packages.
Maybe it would better if you create a new virtual environment, say, `venv-transformers-pr`, and switch to it.

> 
> I think I should fetch upstream before that and rebase.

You can try it. I always try to have (very) recent commit from master to work on a new PR. Hope the rebase is smooth in your case here.

sayakpaul committed:
fix: formatting and styling.

sayakpaul commented on src/transformers/models/vit_mae/modeling_vit_mae.py:
```diff
@@ -223,15 +223,18 @@ def __init__(self, config):
         # fixed sin-cos embedding
         self.position_embeddings = nn.Parameter(
-            torch.zeros(1, self.num_patches + 1, config.hidden_size), requires_grad=False
+            torch.zeros(1, self.num_patches + 1, config.hidden_size),
+            requires_grad=False,
```
@ydshieh I am unable to do the rebasing as mentioned in https://github.com/huggingface/transformers/pull/16255#issuecomment-1072669273 as @ariG23498 is the repo owner. 

Here's what I did before committing the changes:

* I cloned https://github.com/huggingface/transformers in a separate directory.
* Initialized a new Python virtual environment. 
* `cd`'d to this cloned `transformers` and rand `pip install -e .[quality]`. 
* After that from the forked `transformers` directory (from the `tf-mae` branch) I ran `make style`. 

Let me know if these steps are good to go.

ydshieh replied:
I am also new to this usecase. It seems logical and the effects are there. Therefore I think it is OK.

sayakpaul committed:
fix: formatting and removed the copy messages to prevent build fail.

sayakpaul committed:
chore: resolved conflict.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_16255). All of your documentation changes will be reflected on that endpoint.

sayakpaul committed:
fix: styling and formatting based on the PR changes.

sayakpaul committed:
fix: flake8 errors due to whitespaces in the blanklines.

sayakpaul committed:
fix: whitespaces in the blankline.

sayakpaul committed:
removed unnecessary files from the previous commit.

sayakpaul committed:
chore: removed noted about positional embedding prototype.

sayakpaul committed:
fix: styling for the tf vit-mae script.

ariG23498 committed:
Merge branch 'master' into tf-mae
OK

ariG23498 committed:
chore: adding partial testing

ariG23498 committed:
chore: partial fix for tests

ariG23498 committed:
chore: partial test fix

ariG23498 committed:
chore: partial fix with training parameter

ariG23498 committed:
chore: make style

gante commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1167 @@
+
+        super().build(input_shape)
+
+    def random_masking(self, sequence: tf.Tensor, noise: tf.Tensor = None):
```
```suggestion
    def random_masking(self, sequence: tf.Tensor, noise: Optional[tf.Tensor] = None):
```

gante commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1167 @@
+
+        Args:
+            sequence (`tf.Tensor` of shape `(batch_size, sequence_length, dim)`)
+            noise (`tf.Tensor` of shape `(batch_size, sequence_length)`) which is
```
```suggestion
            noise (`tf.Tensor` of shape `(batch_size, sequence_length)`, *optional*) which is
```

gante commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1167 @@
+        training: bool = False,
+        **kwargs,
+    ) -> Union[TFViTMAEModelOutput, Tuple[tf.Tensor]]:
+        inputs = input_processing(
+            func=self.call,
+            config=self.config,
+            input_ids=pixel_values,
+            head_mask=head_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            training=training,
+            kwargs_call=kwargs,
+        )
```
We can now get rid of all `input_processing` calls, as well as the use of all `inputs["foo"]`, with the use of a new decorator, `@unpack_inputs`.

This issue (https://github.com/huggingface/transformers/issues/16051) holds a call for contributions to apply the decorator, as well as the instructions to update the code. It'd be great if the PR used it 🙏

gante reviewed (commented):
Had a diagonal look at the PR, looks cool! LMK if you need help with any outstanding issue :)

gante commented on src/transformers/models/vit_mae/modeling_vit_mae.py:
```diff
@@ -240,18 +240,21 @@ def initialize_weights(self):
 
         Args:
             sequence (`torch.LongTensor` of shape `(batch_size, sequence_length, dim)`)
+            noise (`torch.FloatTensor` of shape `(batch_size, sequence_length)`) which is
```
```suggestion
            noise (`torch.FloatTensor` of shape `(batch_size, sequence_length)`, *optional*) which is
```

sayakpaul committed:
fix: test cases (2 cases to be fixed.

sayakpaul committed:
feat: using unpack_inputs to simplify input_processing.

sayakpaul committed:
chore: removed unneeded import.

sayakpaul committed:
chore: style changes.

sayakpaul committed:
chore: styling change for tf vit mae test.

sayakpaul committed:
chore: removed unused import.

ariG23498 committed:
Merge branch 'master' into tf-mae
OK

ariG23498 committed:
chore: adding docs and auto

ariG23498 committed:
fix: typo in vit_mae source doc

sayakpaul commented on docs/source/model_doc/vit_mae.mdx:
```diff
@@ -47,7 +47,8 @@ alt="drawing" width="600"/>
 <small> MAE architecture. Taken from the <a href="https://arxiv.org/abs/2111.06377">original paper.</a> </small>
 
-This model was contributed by [nielsr](https://huggingface.co/nielsr). The original code can be found [here](https://github.com/facebookresearch/mae).
+This model was contributed by [nielsr](https://huggingface.co/nielsr). TensorFlow version of the model was contributed by [sayakpaul](https://github.com/sayakpaul) and 
+[ariG23498](https://github.com/ariG23498) (equal contribution). The original code can be found [here](https://github.com/facebookresearch/mae). 
```
We could also maybe link [our tutorial](https://keras.io/examples/vision/masked_image_modeling/) here and say something like:

> For a visual understanding of how MAEs work you can check out [this post](https://keras.io/examples/vision/masked_image_modeling/).

ariG23498 replied:
Great idea! Will get to this.

sayakpaul replied:
Cool. Including it under the **Tips** section would be the best IMO.

sayakpaul commented on src/transformers/__init__.py:
```diff
@@ -2094,6 +2094,14 @@
+    _import_structure["models.vit_mae"].extend(
+        [
+            "TFViTMAEForPreTraining",
+            "TFViTMAELayer",
```
I don't this needs to be added.

ariG23498 committed:
chore: mention keras blogpost

sayakpaul commented on src/transformers/__init__.py:
```diff
@@ -4094,6 +4102,7 @@
         )
         from .models.vision_encoder_decoder import TFVisionEncoderDecoderModel
         from .models.vit import TFViTForImageClassification, TFViTModel, TFViTPreTrainedModel
+        from .models.vit_mae import TFViTMAEForPreTraining, TFViTMAELayer, TFViTMAEModel, TFViTMAEPreTrainedModel
```
Same here. No need to import `TFViTMAELayer` I think.

sayakpaul commented on src/transformers/models/vit_mae/__init__.py:
```diff
@@ -33,6 +33,13 @@
         "ViTMAEPreTrainedModel",
     ]
 
```
The import needs to be `from ...file_utils import _LazyModule, is_tf_available, is_torch_available, is_vision_available` if not already done.

sayakpaul commented on src/transformers/models/vit_mae/__init__.py:
```diff
@@ -33,6 +33,13 @@
+if is_tf_available():
+    _import_structure["modeling_tf_vit_mae"] = [
+        "TFViTMAEForPreTraining",
+        "TFViTMAELayer",
```
No main layer.

sayakpaul commented on src/transformers/models/vit_mae/__init__.py:
```diff
@@ -46,6 +53,9 @@
         )
 
+    if is_tf_available():
+        from .modeling_tf_vit_mae import TFViTMAEForPreTraining, TFViTMAELayer, TFViTMAEModel, TFViTMAEPreTrainedModel
```
Same.

sayakpaul replied:
Also didn't see the changes to `src/transformers/utils/dummy_tf_objects.py`.

sayakpaul commented on docs/source/model_doc/vit_mae.mdx:
```diff
@@ -47,7 +47,10 @@ alt="drawing" width="600"/>
+This model was contributed by [nielsr](https://huggingface.co/nielsr). TensorFlow version of the model was contributed by [sayakpaul](https://github.com/sayakpaul) and 
+[ariG23498](https://github.com/ariG23498) (equal contribution). The original code can be found [here](https://github.com/facebookresearch/mae). 
+
+For a visual understanding of how MAEs work you can check out this [Keras post](https://keras.io/examples/vision/masked_image_modeling/).
```
"Keras post" sounds weird to me. I think just "post" is fine.

ariG23498 committed:
chore: removing TFViTMAEMailnLayer

sayakpaul committed:
chore: fix styling for TF vit mae test.

sayakpaul committed:
chore: added rest of the modifications made w.r.t build logs.

sayakpaul committed:
chore: modifications w.r.t build logs.

sayakpaul committed:
fix: test cases and did changes necessary to fit the fixes.

sayakpaul committed:
chore: removed unsused import.

sayakpaul commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        # if tf.math.equal(shape_list(imgs)[1], 3):
+        #     imgs = tf.transpose(imgs, perm=(0, 2, 3, 1))
+
+        imgs = tf.cond(
+            tf.math.equal(shape_list(imgs)[1], 3), lambda: tf.transpose(imgs, perm=(0, 2, 3, 1)), lambda: imgs
```
This was needed to ensure `shape_list()` outputs were compatible in graph and eager modes.

sayakpaul commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        )
+
+        p = self.vit.embeddings.patch_embeddings.patch_size[0]
+        tf.debugging.assert_equal(shape_list(imgs)[1], shape_list(imgs)[2])
+        tf.debugging.assert_equal(shape_list(imgs)[1] % p, 0)
```
Same as above.

sayakpaul commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        """
+        p = self.vit.embeddings.patch_embeddings.patch_size[0]
+        h = w = int(shape_list(x)[1] ** 0.5)
+        tf.debugging.assert_equal(h * w, shape_list(x)[1])
```
Same as above.

sayakpaul committed:
fix: tf vit-mae test cases.

ariG23498 committed:
Merge branch 'main' into tf-mae
OK

sgugger commented on docs/source/index.mdx:
```diff
@@ -259,7 +259,7 @@ Flax), PyTorch, and/or TensorFlow.
 |         VisualBert          |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
 |             ViT             |       ❌       |       ❌       |       ✅        |         ✅         |      ✅      |
-|           ViTMAE            |       ❌       |       ❌       |       ✅        |         ❌         |      ❌      |
+|           ViTMAE            |       ❌       |       ❌       |       ✅        |         ✅         |      ❌      |
```
Nice!

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+
+    omega = tf.range(embed_dim // 2, dtype="float32")
+    omega /= embed_dim / 2.0
+    omega = 1.0 / 10000**omega  # (D/2,)
+
+    pos = tf.reshape(pos, [-1])  # (M,)
+    out = tf.einsum("m,d->md", pos, omega)  # (M, D/2), outer product
+
+    emb_sin = tf.sin(out)  # (M, D/2)
+    emb_cos = tf.cos(out)  # (M, D/2)
+
+    emb = tf.concat([emb_sin, emb_cos], axis=1)  # (M, D)
```
I would expand a bit those comments to use a longer (and more understandable) name, otherwise they aren't very useful.

(I know they are copied from the PyTorch side, but that's an overlook on the review of that modeling file.)

sayakpaul replied:
Agreed. Updated accordingly.

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        # Reshape from [batch_size, seq_length, all_head_size] to [batch_size, seq_length, num_attention_heads, attention_head_size]
+        tensor = tf.reshape(
+            tensor=tensor,
+            shape=(
+                batch_size,
+                -1,
+                self.num_attention_heads,
+                self.attention_head_size,
+            ),
```
```suggestion
            shape=(batch_size, -1, self.num_attention_heads, self.attention_head_size),
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        )
+        self.dropout = tf.keras.layers.Dropout(rate=config.hidden_dropout_prob, name="dropout")
+
+    def call(
+        self,
+        hidden_states: tf.Tensor,
+        input_tensor: tf.Tensor,
+        training: bool = False,
+    ) -> tf.Tensor:
```
```suggestion
    def call(self, hidden_states: tf.Tensor, input_tensor: tf.Tensor, training: bool = False) -> tf.Tensor:
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+            output_attentions=output_attentions,
+            training=training,
+        )
+        attention_output = self.dense_output(
+            hidden_states=self_outputs[0],
+            input_tensor=input_tensor,
+            training=training,
+        )
```
```suggestion
        attention_output = self.dense_output(
            hidden_states=self_outputs[0], input_tensor=input_tensor, training=training
        )
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+            kernel_initializer=get_initializer(config.initializer_range),
+            name="dense",
+        )
+        self.dropout = tf.keras.layers.Dropout(
+            rate=config.hidden_dropout_prob,
+            name="dropout",
+        )
```
```suggestion
        self.dropout = tf.keras.layers.Dropout(rate=config.hidden_dropout_prob, name="dropout")
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        intermediate_output = self.intermediate(hidden_states=layer_output)
+
+        # second residual connection is done here
+        layer_output = self.vit_output(
+            hidden_states=intermediate_output,
+            input_tensor=hidden_states,
+            training=training,
+        )
```
```suggestion
        layer_output = self.vit_output(
            hidden_states=intermediate_output, input_tensor=hidden_states, training=training
        )
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        **kwargs,
+    ) -> Union[TFViTMAEModelOutput, Tuple[tf.Tensor]]:
+        embedding_output, mask, ids_restore = self.embeddings(
+            pixel_values=pixel_values,
+            training=training,
+            noise=noise,
```
```suggestion
            pixel_values=pixel_values, training=training, noise=noise
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+            `Dict[str, tf.Tensor]`: The dummy inputs.
+        """
+        VISION_DUMMY_INPUTS = tf.random.uniform(
+            shape=(
+                3,
+                self.config.num_channels,
+                self.config.image_size,
+                self.config.image_size,
+            ),
```
```suggestion
            shape=(3,  self.config.num_channels, self.config.image_size, self.config.image_size),
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+    )
+    def serving(self, inputs):
+        """
+        Args:
+        Method used for serving the model.
```
```suggestion
        Method used for serving the model.

        Args:
```

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        """
+        imgs: (N, H, W, 3) x: (N, L, patch_size**2 *3)
+        """
+        # if tf.math.equal(shape_list(imgs)[1], 3):
+        #     imgs = tf.transpose(imgs, perm=(0, 2, 3, 1))
```
To clean up?

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+
+    def unpatchify(self, x):
+        """
+        x: (N, L, patch_size**2 *3) imgs: (N, H, W, 3)
```
Same comment as above.

sgugger commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1133 @@
+        ids_restore = outputs.ids_restore
+        mask = outputs.mask
+
+        decoder_outputs = self.decoder(latent, ids_restore)  # [N, L, p*p*3]
```
This comment would be better with real readable names :-)

sgugger commented on tests/vit_mae/test_modeling_vit_mae.py:
```diff
@@ -139,11 +139,7 @@ def create_and_check_for_pretraining(self, config, pixel_values, labels):
-            pixel_values,
-            labels,
-        ) = config_and_inputs
+        (config, pixel_values, labels) = config_and_inputs
```
```suggestion
        config, pixel_values, labels = config_and_inputs
```

sgugger reviewed (approved):
Thanks a lot for your PR! The PR is very clean, I just left some styling/documentation nits.

I really like the approach taken to test that the random part of the model produces the same results in PyTorch and TensorFlow. I think it would be great to refactor it in the common tester at some point (in a followup PR, @ydshieh ) as I know some other models deactivated those tests instead of dealing with the problem.

ydshieh commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,552 @@
+        self.model_tester.create_and_check_for_pretraining(*config_and_inputs)
+
+    # Since TFViTMAEForPretraining has random masking, we need to fix the noise
+    # to generate masks during test
```
Would like to keep the format

```
# overwrite from common since ...
```
as done in other test scripts :-)

ydshieh commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,552 @@
+            self.assertLess(np.sum(np.abs(output_dict - output_keywords)), 1e-6)
+
+    # Since TFViTMAEForPretraining has random masking, we need to fix the noise
+    # to generate masks during test
```
same as above

ydshieh commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,552 @@
+                self.assert_outputs_same(after_outputs, outputs)
+
+    # Since TFViTMAEForPretraining has random masking, we need to fix the noise
+    # to generate masks during test
```
same as above

ydshieh commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,552 @@
+                    out_1[np.isnan(out_1)] = 0
+
+                max_diff = np.amax(np.abs(out_1 - out_2))
+                self.assertLessEqual(max_diff, 1e-5)
```
I am wondering why you need to set `out_2` and `out_1` manually here, and not just do the same thing as in the tf common file [here](https://github.com/ydshieh/transformers/blob/01485ceec3d2e0a9a957ec86f0a10096cecb4a94/tests/test_modeling_tf_common.py#L154).

Of course, the part dealing with the randomness is necessary, and I am fine with that part.

sayakpaul replied:
Took from the PT test. Cc @NielsRogge

ydshieh replied:
OK, you can leave it as it is for this PR :-) Thanks.

ydshieh commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,552 @@
+        reason="""ViTMAE returns a random mask + ids_restore in each forward pass. See test_save_load
+    to get deterministic results."""
+    )
+    def test_save_load_fast_init_from_base(self):
```
As far as I know, this test is only for PT models, and there is no need to add it here (and then skip it)

ydshieh replied:
Although I am quite confident, let @sgugger double-check here :-)

ydshieh replied:
maybe @sayakpaul you mean to put this into the PT test file ?

sayakpaul replied:
Yup. You're right. The test should not be in the TF test.

ydshieh commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,552 @@
+        reason="""ViTMAE returns a random mask + ids_restore in each forward pass. See test_save_load
+    to get deterministic results."""
+    )
+    def test_save_load_fast_init_to_base(self):
```
only for PT

sayakpaul committed:
chore: addressed pr comments.

sayakpaul committed:
Merge branch 'tf-mae' of https://github.com/ariG23498/transformers into tf-mae

sayakpaul committed:
chore: styling fix for common and resolved conflict.

sayakpaul commented:
@sgugger @ydshieh thank you for your positive comments. In the world of open-source, we couldn't have asked for me. So, thank you once again. 

We've addressed the PR comments. Let us know your thoughts.

ariG23498 committed:
fix: typo

sayakpaul commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,582 @@
+        # make random mask reproducible across the PT and TF model
+        np.random.seed(2)
+
+        model = TFViTMAEForPreTraining.from_pretrained(
+            "facebook/vit-mae-base", from_pt=True
+        )  # remove `from_pt` when TF weights have been uploaded.
```
@gante we'd need your help here.

sayakpaul replied:
@gante a gentle reminder about this.

ariG23498 committed:
Merge branch 'main' into tf-mae
OK

sgugger commented:
On top of the code quality issue flagged above, the two following tests are failing:
```
FAILED tests/vit_mae/test_modeling_vit_mae.py::ViTMAEModelTest::test_pt_tf_model_equivalence
FAILED tests/vit_mae/test_modeling_tf_vit_mae.py::TFViTMAEModelTest::test_pt_tf_model_equivalence
```

Could you fix them so we can merge the PR? Thanks a lot!

sayakpaul committed:
chore: removed unnecessary variable.

sayakpaul commented:
> On top of the code quality issue flagged above, the two following tests are failing:
> 
> ```
> FAILED tests/vit_mae/test_modeling_vit_mae.py::ViTMAEModelTest::test_pt_tf_model_equivalence
> FAILED tests/vit_mae/test_modeling_tf_vit_mae.py::TFViTMAEModelTest::test_pt_tf_model_equivalence
> ```
> 
> Could you fix them so we can merge the PR? Thanks a lot!

I looked into it but couldn't figure out why they would fail. Could you provide us with a starting point to get them fixed?

sgugger commented:
Here is the full stack trace:
```
============================= FAILURES SHORT STACK =============================
________________ TFViTMAEModelTest.test_pt_tf_model_equivalence ________________


tf_model = <transformers.models.vit_mae.modeling_tf_vit_mae.TFViTMAEModel object at 0x7f8bef271450>
pt_state_dict = OrderedDict([('embeddings.cls_token', tensor([[[ 0.0355,  0.0288,  0.0095,  0.0219, -0.0116,  0.0049,  0.0247,
       ...0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0.]))])
tf_inputs = {'pixel_values': <tf.Tensor: shape=(13, 3, 30, 30), dtype=float32, numpy=
array([[[[9.84787583e-01, 7.89411783e-01, 8....20547986e-01, 9.40254509e-01, ...,
          1.77839562e-01, 9.66387928e-01, 8.72015655e-01]]]],
      dtype=float32)>}
allow_missing_keys = False

    def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, allow_missing_keys=False):
        """Load pytorch state_dict in a TF 2.0 model."""
        try:
            import tensorflow as tf  # noqa: F401
            import torch  # noqa: F401
            from tensorflow.python.keras import backend as K
        except ImportError:
            logger.error(
                "Loading a PyTorch model in TensorFlow, requires both PyTorch and TensorFlow to be installed. Please see "
                "https://pytorch.org/ and https://www.tensorflow.org/install/ for installation instructions."
            )
            raise
    
        if tf_inputs is None:
            tf_inputs = tf_model.dummy_inputs
    
        if tf_inputs is not None:
            tf_model(tf_inputs, training=False)  # Make sure model is built
        # Adapt state dict - TODO remove this and update the AWS weights files instead
        # Convert old format to new format if needed from a PyTorch state_dict
        old_keys = []
        new_keys = []
        for key in pt_state_dict.keys():
            new_key = None
            if "gamma" in key:
                new_key = key.replace("gamma", "weight")
            if "beta" in key:
                new_key = key.replace("beta", "bias")
            if new_key:
                old_keys.append(key)
                new_keys.append(new_key)
        for old_key, new_key in zip(old_keys, new_keys):
            pt_state_dict[new_key] = pt_state_dict.pop(old_key)
    
        # Make sure we are able to load PyTorch base models as well as derived models (with heads)
        # TF models always have a prefix, some of PyTorch models (base ones) don't
        start_prefix_to_remove = ""
        if not any(s.startswith(tf_model.base_model_prefix) for s in pt_state_dict.keys()):
            start_prefix_to_remove = tf_model.base_model_prefix + "."
    
        symbolic_weights = tf_model.trainable_weights + tf_model.non_trainable_weights
        tf_loaded_numel = 0
        weight_value_tuples = []
        all_pytorch_weights = set(list(pt_state_dict.keys()))
        missing_keys = []
        for symbolic_weight in symbolic_weights:
            sw_name = symbolic_weight.name
            name, transpose = convert_tf_weight_name_to_pt_weight_name(
                sw_name, start_prefix_to_remove=start_prefix_to_remove, tf_weight_shape=symbolic_weight.shape
            )
    
            # Find associated numpy array in pytorch model state dict
            if name not in pt_state_dict:
                if allow_missing_keys:
                    missing_keys.append(name)
                    continue
                elif tf_model._keys_to_ignore_on_load_missing is not None:
                    # authorized missing keys don't have to be loaded
                    if any(re.search(pat, name) is not None for pat in tf_model._keys_to_ignore_on_load_missing):
                        continue
>               raise AttributeError(f"{name} not found in PyTorch model")
E               AttributeError: vit.embeddings.cls_token not found in PyTorch model

src/transformers/modeling_tf_pytorch_utils.py:198: AttributeError
_________________ ViTMAEModelTest.test_pt_tf_model_equivalence _________________


tf_model = <transformers.models.vit_mae.modeling_tf_vit_mae.TFViTMAEModel object at 0x7f8c0e80add0>
pt_state_dict = OrderedDict([('embeddings.cls_token', tensor([[[-0.0341, -0.0326,  0.0065, -0.0189,  0.0114,  0.0198,  0.0247,
       ...0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0.]))])
tf_inputs = {'pixel_values': <tf.Tensor: shape=(13, 3, 30, 30), dtype=float32, numpy=
array([[[[0.00428793, 0.00807637, 0.5928316 ...,
         [0.0279441 , 0.08120351, 0.30608034, ..., 0.777531  ,
          0.6238623 , 0.38510746]]]], dtype=float32)>}
allow_missing_keys = False

    def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, allow_missing_keys=False):
        """Load pytorch state_dict in a TF 2.0 model."""
        try:
            import tensorflow as tf  # noqa: F401
            import torch  # noqa: F401
            from tensorflow.python.keras import backend as K
        except ImportError:
            logger.error(
                "Loading a PyTorch model in TensorFlow, requires both PyTorch and TensorFlow to be installed. Please see "
                "https://pytorch.org/ and https://www.tensorflow.org/install/ for installation instructions."
            )
            raise
    
        if tf_inputs is None:
            tf_inputs = tf_model.dummy_inputs
    
        if tf_inputs is not None:
            tf_model(tf_inputs, training=False)  # Make sure model is built
        # Adapt state dict - TODO remove this and update the AWS weights files instead
        # Convert old format to new format if needed from a PyTorch state_dict
        old_keys = []
        new_keys = []
        for key in pt_state_dict.keys():
            new_key = None
            if "gamma" in key:
                new_key = key.replace("gamma", "weight")
            if "beta" in key:
                new_key = key.replace("beta", "bias")
            if new_key:
                old_keys.append(key)
                new_keys.append(new_key)
        for old_key, new_key in zip(old_keys, new_keys):
            pt_state_dict[new_key] = pt_state_dict.pop(old_key)
    
        # Make sure we are able to load PyTorch base models as well as derived models (with heads)
        # TF models always have a prefix, some of PyTorch models (base ones) don't
        start_prefix_to_remove = ""
        if not any(s.startswith(tf_model.base_model_prefix) for s in pt_state_dict.keys()):
            start_prefix_to_remove = tf_model.base_model_prefix + "."
    
        symbolic_weights = tf_model.trainable_weights + tf_model.non_trainable_weights
        tf_loaded_numel = 0
        weight_value_tuples = []
        all_pytorch_weights = set(list(pt_state_dict.keys()))
        missing_keys = []
        for symbolic_weight in symbolic_weights:
            sw_name = symbolic_weight.name
            name, transpose = convert_tf_weight_name_to_pt_weight_name(
                sw_name, start_prefix_to_remove=start_prefix_to_remove, tf_weight_shape=symbolic_weight.shape
            )
    
            # Find associated numpy array in pytorch model state dict
            if name not in pt_state_dict:
                if allow_missing_keys:
                    missing_keys.append(name)
                    continue
                elif tf_model._keys_to_ignore_on_load_missing is not None:
                    # authorized missing keys don't have to be loaded
                    if any(re.search(pat, name) is not None for pat in tf_model._keys_to_ignore_on_load_missing):
                        continue
>               raise AttributeError(f"{name} not found in PyTorch model")
E               AttributeError: vit.embeddings.cls_token not found in PyTorch model

src/transformers/modeling_tf_pytorch_utils.py:198: AttributeError
```
Looks like there is a weight present in the TF model but not the PyTorch one.

sayakpaul commented:
This does seem weird. I think if the TF model didn't have the right variable names as same as the PT one, it would not have loaded the PT weights back into the TF model, no?

sayakpaul commented:
@sgugger there seems to be something weird here. I did the following:

```py
from transformers import ViTMAEForPreTraining
model = ViTMAEForPreTraining.from_pretrained("facebook/vit-mae-base")

param_states = model.state_dict()

for k in param_states:
	if "cls" in k:
		print(k)
```

It prints:

```
vit.embeddings.cls_token
```

But as per the error trace it says:

```
AttributeError: vit.embeddings.cls_token not found in PyTorch model
```

ydshieh commented:
Considering it was me working on this PT/TF equivalence test recently, I will try to take a look!

ydshieh commented:
@sayakpaul Could you try if the following work 🙏?

```
TFViTMAEModel.from_pretrained("facebook/vit-mae-base", from_pt=True)
```

ariG23498 committed:
Merge branch 'main' into tf-mae
OK

ydshieh commented:
@sayakpaul 

change [here] from 
```
base_model_prefix = "vit_mae"
```
to
```
base_model_prefix = "vit"
```
will fix the loading issue, because that is the value used in the `ViTMAEModel`, see [here](https://github.com/huggingface/transformers/blob/2da1f03f93af37480fdfc23cbc870f5bf038c478/src/transformers/models/vit_mae/modeling_vit_mae.py#L585).

However I still get the large difference value between PT/TF
```
E   AssertionError: 3.805268 not less than or equal to 1e-05
```
So something still need to be fixed.

@sgugger @NielsRogge Do you think 

- `vit_mae` is a better value for `base_model_prefix`
- if so, should we & are we able to fix this in PyTorch `ViTMAEModel`
- (I am completely fine if we don't change it because we can't)

sayakpaul committed:
fix: base_model_prefix.

sayakpaul commented:
@ydshieh thanks much for pointing us in the right direction. As for the assertion, it likely can be solved by fixing the `noise` argument like the way we did for the rest of the tests. 

If you could suggest where the `noise` argument needs to be added, I can look into it.

ydshieh commented:
@sayakpaul Wow, super! `noise` is something we should keep in mind from now :-)

You can copy/paste the `test_pt_tf_model_equivalence` from `test_modeling_tf_common.py`, then add the `noise` part.
There are a few places we can remove (they only make sense in the common test), but if you are not sure which part to remove, you can keep it as it is - I will make the change.

sayakpaul commented:
Thank you @ydshieh. Making the changes right away.

sayakpaul committed:
fix: pt tf model equivalence.

sayakpaul committed:
fix: equivalence test for the PT model.

sayakpaul committed:
fix: import error and np.float32 assignment error.

sayakpaul commented:
Seems like GitHub Actions is acting weird. But that's not what GitHub's status page says: https://www.githubstatus.com/.

Let's see if the current change can fix the equivalence test.

ydshieh commented:
> Seems like GitHub Actions is acting weird. But that's not what GitHub's status page says: https://www.githubstatus.com/.
> 
> Let's see if the current change can fix the equivalence test.

I have exactly the same situation yesterday - 0 check after force push

sayakpaul commented:
I see. If possible it'd be great if you could review the changes I made so far.

sayakpaul commented:
@sgugger is it possible to trigger the workflows manually from your end?

sgugger commented:
Not from side, no. Can you do an empty commit (`git commit -m "Trigger tests" --low-empty`)?

sayakpaul committed:
chore: empty commit to trigger tests.

sayakpaul commented:
@sgugger seems like things are looking good. Let me know if the recent changes to fix equivalence look good to you. 

This is one last bit that I think you folks need to take care of: https://github.com/huggingface/transformers/pull/16255#discussion_r833968556.

ydshieh commented:
> @sgugger seems like things are looking good. Let me know if the recent changes to fix equivalence look good to you.
> 
> This is one last bit that I think you folks need to take care of: [#16255 (comment)](https://github.com/huggingface/transformers/pull/16255#discussion_r833968556).

Glad to see the tests being all green!

@gante Before merge, please wait for me to remove some irrelevant parts in the overwritten PT/TF equivalence test, thanks!

ydshieh committed:
remove irrelevant blocks

ydshieh commented on tests/vit_mae/test_modeling_vit_mae.py:
```diff
@@ -319,6 +319,8 @@ def check_hidden_states_output(inputs_dict, config, model_class):
             check_hidden_states_output(inputs_dict, config, model_class)
 
+    # overwrite from common since ViTMAEForPretraining has random masking, we need to fix the noise
+    # to generate masks during test
```
just a nit

sayakpaul replied:
Kept it intentionally to explain the purpose of the override.

ydshieh commented:
I have pushed a commit to remove irrelevant blocks (that only make sense in the common test) + put pytorch inputs on the correct device (for GPU testing to run)

ydshieh commented:
> Thanks for your help, @ydshieh. The failing tests seem to be unrelated to the PR?

Yeah, I think it is another model `facebook/xglm-564M` causing issue.
You can have a bit of rest for now 💯 💖

gante reviewed (approved):
Awesome work 🔥 I have a few minor questions, but the PR is almost good to go -- pre-aproving.

As mentioned below, will take care of adding the TF weights.

ariG23498 commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1101 @@
+        return sequence_masked, mask, ids_restore
+
+    def call(self, pixel_values: tf.Tensor, noise: tf.Tensor = None) -> tf.Tensor:
+        batch_size, num_channels, height, width = shape_list(pixel_values)
```
We agree. We had kept this in to be aligned with the PT code.

I will go ahead an remove this!

ariG23498 commented on tests/vit_mae/test_modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,717 @@
+
+            check_pt_tf_models(tf_model, pt_model)
+
+    def test_compile_tf_model(self):
```
I will defer this to @sayakpaul.

ariG23498 commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -0,0 +1,1101 @@
+        return outputs
+
+
+class TFViTMAESelfOutput(tf.keras.layers.Layer):
```
This is a good catch! I will quickly add the references.

ariG23498 committed:
chore: minor review edit

ariG23498 committed:
Merge branch 'main' into tf-mae
OK

ariG23498 commented:
@gante according to https://github.com/huggingface/transformers/pull/16255/commits/debad62d47f5789f9890d3176713345e3a2506c0 there were no key additions or deletions from the code base. Do you know the possible reason for the tests to fail?

ariG23498 committed:
chore: fix typo

ydshieh commented on src/transformers/models/vit_mae/modeling_tf_vit_mae.py:
```diff
@@ -490,6 +492,7 @@ def call(
         return outputs
 
 
+# Copied from transformers.models.vit.modeling_tf_vit.ViTIntermediate with TFViT->TFViTMAE
```
You acted faster than my comment !

ariG23498 replied:
Yep 😅 
Fixed it! Thanks.

Did not know that a test was kept to check for the comments! Absolutely lovely.

ariG23498 replied:
This fix still does not make it I think.

There are failed tests. Could you please guide me here?

ydshieh replied:
For the `check_repository_consistency`, errors like  

```python
- src/transformers/models/vit_mae/modeling_tf_vit_mae.py: copy does not match models.vit.modeling_tf_vit.TFViTSelfAttention at line 355
- src/transformers/models/vit_mae/modeling_tf_vit_mae.py: copy does not match models.vit.modeling_tf_vit.TFViTSelfOutput at line 440
```

means you put a comment stating it is a copy from another place, but actually they are different code block.
I saw gante only made a single comment in the review mentioning to add `# Copied from`, but it seems you added quite a few. Any reason?

ydshieh replied:
However, I do see in `ViTMAE`, there are quite a few `# copied from`. For TensorFlow version `TFViT` and `TFViTMAE`, I don't know if they are different and couldn't use `# copied from`.

ariG23498 replied:
In the [comment](https://github.com/huggingface/transformers/pull/16255#discussion_r835516046) as I understood it Gante wanted to add the comment in all the code snippets taken from the `ViT` model.

This was my only rationale.

ydshieh replied:
I saw you add config type, great.

However, the copy statement you added were `with TFViT->TFViTMAE`, but the added type `ViTMAEConfig` vs `ViTConfig`, and therefore they are not taken into account.

I didn't verify the code blocks, but it is probably could be solved by change `with TFViT->TFViTMAE` to `with ViT->ViTMAE`.

Otherwise, another way is to use `with TFViT->TFViTMAE, ViTConfig->ViTMAEConfig`

ydshieh replied:
It would be easier if you run `make fix-copies` locally to see if the changes would work. (don't forget to stage the changed files  first)

ariG23498 committed:
chore: adding type hints for config argument

sayakpaul committed:
chore: addressed PR comments and attempt at build fixes.

sayakpaul committed:
chore: ran make fix-copies.

sayakpaul committed:
fix: vitconfig -> vitmaeconfig.

sayakpaul commented:
@gante @ydshieh 

Updates:

Repo consistency tells us that copied components have inconsistencies. Sorry, if this sounds foolish but we could not actually figure out those inconsistencies. Class name changes and other relevant changes in the comment do not count here I hope. To ensure the copying was indeed right we went ahead used a comparator tool to verify if that's the case. Here are the individual links to the comparison results for the components that are said to have copy inconsistencies:

* SelfAttention: https://www.diffchecker.com/m5VY19Xl
* SelfOutput: https://www.diffchecker.com/j6Pri9fN
* MAEAttention: https://www.diffchecker.com/5qtn9IRh
* Intermediate: https://www.diffchecker.com/hNy0Sg7c
* ViTOutput: https://www.diffchecker.com/CyK4zZUl
* Layer: https://www.diffchecker.com/BESXzSTD
* Encoder: https://www.diffchecker.com/BXHYohx1

Note that we did run `make fix-copies` within the environment as described in https://github.com/huggingface/transformers/pull/16255#discussion_r830432539 (refer to https://github.com/huggingface/transformers/pull/16255/commits/565ec4c1227d9577e0e4e7489880dad2d4c33d01) and it replaced `ViTMAEConfig` with `ViTConfig` which is wrong, I guess.

What else are we missing out here?

ydshieh commented:
@ariG23498 @sayakpaul 

Probably my previous comment didn't explain things clear enough.

Let's take an example with:

```
# Copied from transformers.models.vit.modeling_tf_vit.TFViTSelfAttention with TFViT->TFViTMAE
class TFViTMAESelfAttention(tf.keras.layers.Layer):
    def __init__(self, config: ViTMAEConfig, **kwargs):
```

This `# copied from` will check if the block is a copy from `TFViTSelfAttention` after replaceing `TFViT` by `TFViTMAE` (in the RAM).
However, there is `ViTConfig` in the ViT file, and `ViTMAEConfig` in this block **(note: there is no TF prefix here)**, and the instruction `with TFViT->TFViTMAE` in the comment `# Copied from` won't apply to  `ViTConfig` (from the ViT file), and during the check, it is `ViTConfig` vs `ViTMAEConfig`, and therefore there is a difference.

In order to fix the issue, they are 2 potential options:

1. In `# Copied from`, change ` with TFViT->TFViTMAE` to ` with ViT->ViTMAE`
  - This will only work if the 2 blocks in ViT and ViTMAE are indeed the same after this replace!

2. In `# Copied from`, change ` with TFViT->TFViTMAE` to ` with TFViT->TFViTMAE, ViTConfig->ViTMAEConfig`
  - This is more verbose.

So I would suggest try option 1 first. If there are remaining places, we can use option 2 for that remaining places.

ariG23498 committed:
chore: comment fix ViT->ViTMAE

ydshieh commented:
For the remaining # copped issue, 

```
# in ViT, layernorm is applied before self-attention
```
to
```
# in ViTMAE, layernorm is applied before self-attention
```
should work.
(This is also what has done in PyTorch ViTMAE)

ariG23498 committed:
chore: comment fix in TFViTMAELayer

ydshieh commented:
It turns out that there is still 

```
# in ViT, layernorm is also applied after self-attention
```
to be changed ...I didn't check the whole block previously

ariG23498 committed:
chore: fix comment in TFViTMAELayer

ariG23498 commented:
Thanks for the prompt feedback @ydshieh

It was my bad to not check the entire code block.

ydshieh commented:
Don't worry. 

And just a **potential** correction (not really important): my previous comment "it would be easier to detect ..." **might** be not True in this particular line: I don't think there will be a diff visible to this particular line, after running `make fix-copies`.

ariG23498 commented:
All green ✅ 

Thanks @ydshieh for your prompt and valuable feedback!

ydshieh commented on tests/vit_mae/test_modeling_vit_mae.py:
```diff
@@ -400,34 +544,31 @@ def default_feature_extractor(self):
-        # source: https://discuss.pytorch.org/t/random-seed-that-spans-across-devices/19735
-        torch.manual_seed(2)
+        # make random mask reproducible across the PT and TF model
+        np.random.seed(2)
```
For this change, it would be great to hear from @NielsRogge too.

Previously we only saw the issue caused by CPU/GPU PRGNs (for the PyTorch ViTMAE).
But now we also see the issue caused by the different frameworks.

I think using `numpy` is a good choice to address this issue. However, let's wait Niels' opinion.

sayakpaul replied:
It was discussed with @NielsRogge beforehand, though. 

But I don't think this PR is contingent on this change as it was all well discussed before and reviewed as well. I might wrong though.

ydshieh replied:
I know we discussed on Slack. As far as I know, Niels knows we are going to use the new `noise` argument and he is happy with this change.

However, regarding using `numpy` instead of `torch` (and updating the expected slice), I don't remember if Niels is aware of this change and expressed his opinion. I might be wrong (?).

Regarding `But now we also see the issue caused by the different frameworks.`, it is just about we decided to use `numpy` because the PRNGs from torch and tf will give different results with the same seed, and we don't want to import torch in tf (non pt-tf cross) test.

sayakpaul replied:
Thanks for clarifying! 

> because the PRNGs from torch and tf will give different results even with the same seed, 

I think you probably meant TF PT PRNGs won't produce same results even if they were seeded with the same number?

ydshieh replied:
yes, that's what I mean.

ydshieh replied:
Once Niels is OK with using NumPy, I think it is better to have the comment like

```
# make random mask reproducible across different frameworks (PyTorch, TensorFlow, etc.), as well as across different accelerators (CPU, GPU, etc.).
# note that the same seed on CPU and on GPU doesn’t mean they spew the same random number sequences,
# as they both have fairly different PRNGs (for efficiency reasons).
# source (for PyTorch): https://discuss.pytorch.org/t/random-seed-that-spans-across-devices/19735
```

sayakpaul replied:
Much more comprehensive. Thanks for suggesting 👌

ydshieh replied:
So Niels is OK with this change, I will just let you update the comment, @sayakpaul . Thank you!

ydshieh commented:
About `shape_list() is a mess for graph mode. We have a plan to get rid of it, and just use .shape everywhere `

https://github.com/huggingface/transformers/pull/16255#discussion_r835525428

I was a bit worried it will break some testing at this moment, and tested with `shape`. Indeed, the test `test_save_load` will give error at the line `model.save_pretrained(tmpdirname, saved_model=True)` - when calling `autograph_handler` inside. See the full log below.

## Full error log

```python
            with tempfile.TemporaryDirectory() as tmpdirname:
>               model.save_pretrained(tmpdirname, saved_model=True)

tests\vit_mae\test_modeling_tf_vit_mae.py:616: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src\transformers\modeling_tf_utils.py:1418: in save_pretrained
    self.save(saved_model_dir, include_optimizer=False, signatures=self.serving)
..\..\..\..\miniconda3\envs\py39\lib\site-packages\keras\utils\traceback_utils.py:67: in error_handler
    raise e.with_traceback(filtered_tb) from None
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = ({'pixel_values': <tf.Tensor 'pixel_values:0' shape=(None, None, None, None) dtype=float32>},)
kwargs = {}

    def autograph_handler(*args, **kwargs):
      """Calls a converted version of original_func."""
      # TODO(mdan): Push this block higher in tf.function's call stack.
      try:
        return autograph.converted_call(
            original_func,
            args,
            kwargs,
            options=autograph.ConversionOptions(
                recursive=True,
                optional_features=autograph_options,
                user_requested=True,
            ))
      except Exception as e:  # pylint:disable=broad-except
        if hasattr(e, "ag_error_metadata"):
>         raise e.ag_error_metadata.to_exception(e)
E         ValueError: in user code:
E         
E             File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\models\vit_mae\modeling_tf_vit_mae.py", line 728, in serving  *
E                 return self.call(inputs)
E             File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\modeling_tf_utils.py", line 816, in run_call_with_unpacked_inputs  *
E                 return func(self, **unpacked_inputs)
E             File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\models\vit_mae\modeling_tf_vit_mae.py", line 1074, in call  *
E                 loss = self.forward_loss(pixel_values, logits, mask)
E             File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\models\vit_mae\modeling_tf_vit_mae.py", line 1007, in forward_loss  *
E                 target = self.patchify(imgs)
E             File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\models\vit_mae\modeling_tf_vit_mae.py", line 975, in patchify  *
E                 imgs = tf.cond(
E         
E             ValueError: Tried to convert 'x' to a tensor and failed. Error: None values not supported.

..\..\..\..\miniconda3\envs\py39\lib\site-packages\tensorflow\python\framework\func_graph.py:1129: ValueError
```

ydshieh commented:
Changing `shape_list` to `shape` will also cause a test failing 

https://github.com/huggingface/transformers/blob/b320d87eceb369ea22d5cd73866499851cb2cca3/tests/utils/test_modeling_tf_core.py#L124-L135

with similar error

```python
E             in user code:
E             
E                 File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\modeling_tf_utils.py", line 816, in run_call_with_unpacked_inputs  *
E                     return func(self, **unpacked_inputs)
E                 File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\models\vit_mae\modeling_tf_vit_mae.py", line 1074, in call  *
E                     loss = self.forward_loss(pixel_values, logits, mask)
E                 File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\models\vit_mae\modeling_tf_vit_mae.py", line 1007, in forward_loss  *
E                     target = self.patchify(imgs)
E                 File "C:\Users\33611\Desktop\Projects\transformers-huggingface\transformers\src\transformers\models\vit_mae\modeling_tf_vit_mae.py", line 980, in patchify  *
E                     tf.debugging.assert_equal(imgs.shape[1], imgs.shape[2])
E             
E                 ValueError: None values not supported.
```


**(This test is currently only used for just a few core NLP models)**

sayakpaul commented:
Thanks for investigating it, @ydshieh. Is there anything we can do in this PR to mitigate the problem?

ydshieh commented:
> Thanks for investigating it, @ydshieh. Is there anything we can do in this PR to mitigate the problem?

Let's see what gante think.

sayakpaul commented:
> https://github.com/huggingface/transformers/pull/16255#issuecomment-1079774153

This is why I used `shape_list()` there. The root cause here is that inside `patchify()` with `.shape` there will be a `None` in the batch size and `tf.random.uniform()` will error out for that. I could not think of any other workaround to mitigate the problem. 

For the second one (https://github.com/huggingface/transformers/pull/16255#issuecomment-1079785554) I am not too sure since you mentioned that the test is only applied for some core NLP models.

ydshieh commented:
> For the second one ([#16255 (comment)](https://github.com/huggingface/transformers/pull/16255#issuecomment-1079785554)) I am not too sure since you mentioned that the test is only applied for some core NLP models.

My comments is not about `we should change to shape and fix all the failing tests`.

It is more about a question (to gante) `if we should make this decision and change in this PR, considering some tests will fail.`
(And the second comment is included mainly for us (HF) not to forget there are some tests we want to run for all models ideally, but currently run only for a few models. We (HF) shouldn't make a big decision and see all tests being green, but will fail some important tests that are not currently run due to some limitation).

sayakpaul commented:
Now i understand. Appreciate the class clarification.

gante commented:
@sayakpaul `vit-mae-base` [TF weights are on the hub](https://huggingface.co/facebook/vit-mae-base/tree/main), the other will soon follow :) I think you can make the final changes now (remove the `from_pt`), to then merge the PR 💪

sayakpaul committed:
chore: removed from_pt argument.

sayakpaul commented:
@gante thank you! Changes made.

gante commented:
btw ignore the failure in `Add model like runner / Add new model like template tests (pull_request)`, it's being looked after

sayakpaul commented:
> > @gante thank you! Changes made.
> 
> Can you confirm that the tests run with `RUN_SLOW=1`? Will merge after CI gets to green and I get your confirmation 🚀

Yes, I did run the tests before pushing the changes.

sayakpaul commented:
@gante over to you to take the reigns.

gante commented:
CI green, slow tests were run, all TF weights on the hub 👉 merging.

Great work @sayakpaul @ariG23498 💪
